Progress

pbrehmer opened the pull request:
- closes #21 
- closes #26 
- closes #20 
- closes #13 
- closes #28 
- closes #24
- closes #19 
- closes #18 
- closes #17 
- closes #27

mfherbst commented on docs/src/examples/xxz_dmrg.md:
```diff
@@ -12,14 +12,11 @@ Instead of constructing explicit matrices from Kronecker products as we did befo
-using ITensors
-using Plots
-using ReducedBasis
+using LinearAlgebra, ITensors, Plots, ReducedBasis
```
Not too big a fan of putting these on one line, acutally. It makes it easy to miss one thing.

pbrehmer replied:
I thought it was okay for the examples, since they have not that many dependencies, but I can change it back to the initial formatting.

mfherbst commented on src/affine_decomposition.jl:
Great idea, but ⪒ is quite tough to read. Pick a better symbol (or just a function name)

mfherbst replied:
how about `is_nonredundant`

pbrehmer replied:
Yes that would be a good name -- earlier I just struggled to find a fitting name, so I thought I would just take the fitting symbol. I'll change that to `is_nonredundant` then.

mfherbst commented on src/affine_decomposition.jl:
isnothing should be directly usable here, no anonymous function needed.

pbrehmer replied:
Ah good to know!

mfherbst commented on src/affine_decomposition.jl:
Why is this needed? Should the `map` not take care of this?

pbrehmer replied:
Unfortunately not, since for the indices that are not computed, the map returns `nothing` so that the `matrixel` matrix is of the type `Matrix{Union{nothing, Matrix{ComplexF64}}}.

mfherbst replied:
Then just don't return nothing, but use zero instead and symmetrise the matrix afterwards (compute A + A')

pbrehmer replied:
Hmm, but the symmetrization A + A' then creates `2*matrix[i, i]` on the diagonal elements but not on the non-diagonal elements, right? So I would have to correct this factor of 2.

pbrehmer replied:
I further tried to get rid of this bit, but couldn't get it to work yet. The problem with setting elements to zeros is that one needs to infer the type of the computed elements (`Matrix{ComplexF64}` in most cases), which is the old problem again. Just returning `0` produces as mixed-type matrix and leads to matrix-integer additions in the A+A' step.

Maybe we should try this together at some later point?

mfherbst replied:
ok. Let's have an issue.

pbrehmer replied:
Yes, I opened #35.

pbrehmer commented:
@mfherbst the things we discussed earlier should now be (correctly) implemented. I also added some comments to related issues #12 and #33.

mfherbst commented on src/affine_decomposition.jl:
Could this special-casing not be included above?

pbrehmer replied:
Hmm I guess I could include this in the `findall(isnothing(x), matrixel)` for-loop, such that the `nothing` elements get explicitly computed when there is no redundancy.

mfherbst replied:
add this to the issue above.

mfherbst commented on src/callback.jl:
```diff
@@ -4,23 +4,22 @@
 Print diagnostic information in each assembly iterations.
 """
 function print_callback(info)
```
This callback is never executed in the tests. Do this somewhere to avoid making a stupid mistake without notice.

pbrehmer replied:
Yes that's a good point. I will do a dedicated test for that.

mfherbst commented on src/truncate.jl:
```diff
@@ -0,0 +1,92 @@
+but the number of truth solves included in the basis, which can feature degenerate
+snapshots.
+"""
+function Base.truncate(basis::RBasis, n_truth::Int)
```
Most of this is not yet tested.

pbrehmer replied:
Yes, I will add some tests tomorrow, today I ran out of time :-)

pbrehmer commented:
I now implemented the suggestions from the comments above, except for the `promote_type.(matrixel)` bit. Aside from the `truncate` and `print_callback` tests, I also added some smaller "obvious" tests throughout, after I had a look at the code coverage data.
